src/ml/neural_net/style_transfer/mps_style_transfer.m

```diff
@@ -0,0 +1,419 @@
+    [contentMeanArray addObject:contentMean];
+
+    MPSImage *contentMultiplication = [[MPSImage alloc] initWithDevice:_dev imageDescriptor:imgDesc];
+    [contentMultiplication writeBytes:multiplication.bytes dataLayout:(MPSDataLayoutHeightxWidthxFeatureChannels)imageIndex:0]; 
+    [contentMultiplicationArray addObject:contentMultiplication];
+
+    MPSImage *syleImage = [[MPSImage alloc] initWithDevice:_dev imageDescriptor:imgDesc];
+    [syleImage writeBytes:inputs[styleKey].bytes dataLayout:(MPSDataLayoutHeightxWidthxFeatureChannels)imageIndex:0];
+    [styleImageArray addObject:syleImage];
+
+    MPSImage *styleMean = [[MPSImage alloc] initWithDevice:_dev imageDescriptor:imgDesc];
+    [styleMean writeBytes:mean.bytes dataLayout:(MPSDataLayoutHeightxWidthxFeatureChannels)imageIndex:0];
+    [styleMeanArray addObject:styleMean];
+
+    MPSImage *styleMultiplication = [[MPSImage alloc] initWithDevice:_dev imageDescriptor:imgDesc];
+    [styleMultiplication writeBytes:multiplication.bytes dataLayout:(MPSDataLayoutHeightxWidthxFeatureChannels)imageIndex:0];
+    [styleMultiplicationArray addObject:styleMultiplication];
+  }
+  
+  MPSImageBatch *contentImageBatch = [contentImageArray copy];  
+  MPSImageBatch *contentMeanBatch = [contentMeanArray copy];  
+  MPSImageBatch *contentMultiplicationBatch = [contentMultiplicationArray copy];  
+
+  MPSImageBatch *styleImageBatch = [styleImageArray copy];  
+  MPSImageBatch *styleMeanBatch = [styleMeanArray copy];  
+  MPSImageBatch *styleMultiplicationBatch = [styleMultiplicationArray copy];  
+  
+
+  id<MTLCommandBuffer> cb = [_commandQueue commandBuffer];
+
+  NSMutableArray *intermediateImages = [[NSMutableArray alloc] initWithArray:@[]];
+  NSMutableArray *destinationStates = [[NSMutableArray alloc] initWithArray:@[]];
+
+  MPSImageBatch *ret =  [_trainingGraph encodeBatchToCommandBuffer:cb
+                                                      sourceImages:@[contentImageBatch, contentMultiplicationBatch, contentMeanBatch, styleImageBatch, styleMultiplicationBatch, styleMeanBatch]
+                                                      sourceStates:nil
+                                                intermediateImages:intermediateImages
+                                                 destinationStates:destinationStates];
+
+  for (MPSImage *image in ret) {
+    [image synchronizeOnCommandBuffer:cb];  
+  }
+
+  [cb commit];
+  [cb waitUntilCompleted];
+
+  // AAAK: What about overflow?
+  float lossValue = 0.0f;
+  MPSImageBatch *outBatch = [intermediateImages objectAtIndex:0];
+  for (NSUInteger index = 0; index < _batchSize; index++) {
+    float loss;
+    MPSImage* lossImage = outBatch[index];
+    [lossImage readBytes:&loss dataLayout:(MPSDataLayoutHeightxWidthxFeatureChannels)imageIndex:0];
+    lossValue += loss;
+  }
+  lossValue /= _batchSize;
+
+  NSMutableData * lossData = [NSMutableData dataWithCapacity:0];
```

Comment:
Here too use `dataWithLength:` instead of `dataWithCapacity`.

The length should be `sizeof(float)`, right?

Might be worth factoring out these NSData operations into a helper function or category method.

Reply:
```
NSMutableData * lossData = [NSMutableData dataWithLength:0];
[lossData appendBytes:&lossValue length:sizeof(float)];
```

In this instance I'm appending the bytes so the length being 0 is fine right?

Reply:
Ah, I believe you’re right. Probably simplest to use either `[NSMutableData data]` or `[NSMutableData dataWithCapacity:sizeof(float)]`, though

Reply:
fixed